akka-http-core/src/main/scala/akka/http/impl/engine/client/pool/NewHostConnectionPool.scala

```diff
@@ -102,9 +115,15 @@ private[client] object NewHostConnectionPool {
             } else if (!hasBeenPulled(requestsIn))
               pull(requestsIn)
 
-        def hasIdleSlots: Boolean =
-          // TODO: optimize by keeping track of idle connections?
-          slots.exists(_.isIdle)
+        def hasIdleSlots: Boolean = {
+          if (log.isDebugEnabled) { // somewhat sneaky way of enabling extra assertions in "debug-mode"
+            // Helps debugging if you suspect that idleSlots are not consistent with actual state any more
+            val idle = idleSlots.asScala.map(_.slotId).toSet
+            val idleAll = slots.filter(_.isIdle).map(_.slotId).toSet
+            require(idle == idleAll, s"Managed idle [${idle.mkString(", ")}] != real idle [${idleAll.mkString(", ")}]")
```

Comment:
Should we make a note to eventually remove this check?

Reply:
The question is when. Right now there probably are no bugs. The check helps especially when after you made any changes.

Reply:
Maybe the right way to do would be to create a real option to enable more assertions and enable it when running tests.